networkx/readwrite/text.py

```diff
@@ -0,0 +1,155 @@
+"""
+Text-based visual representations of graphs
+"""
+
+
+def forest_str(graph, with_labels=True, sources=None, write=None):
```

Comment:
This function produces nice visualizations, and I think having visualizations that work in the terminal is a really nice feature.

One thing that I've [run into before](https://github.com/numpy/numpy/pull/15666#issuecomment-601337138) is issues with the default fonts for terminals not supporting the full range of unicode characters. This was particularly a problem with Windows, as the default fonts for popular CLI programs only support the Latin1 character range (or some other limited subset). In the linked PR we ended up doing a platform check and adding an ASCII printing style that was the default for windows platforms.

I'm not sure how big of a concern this is here, but I thought I'd mention it as it'd be nice to make sure this is well-supported on all platforms. Unfortunately this is a difficult thing to catch as tests that check the character codes will pass, but the actual character rendering can still fail.

Reply:
I took a lot of time in selecting the characters to use. The ones I've chosen are fairly well supported, but its definitely something we should double check, or perhaps include a fallback mode. I have a bunch of notes on the topic in the comments, I think I removed them in the latest commit, but I'll dig them up.

Reply:
So the characters I chose are "box-drawing characters" (https://en.wikipedia.org/wiki/Box-drawing_character). 

The main characters I use: `╙`, `─`, `╟`, `├` are part of a DOS subset that is widely supported. 

However, I also use two other characters `╎` (to indicate disconnected trees in the same forest) and `╼` (which I use to indicate direction). I'm not sure if they will cause issues. 

I previously used characters from https://stackoverflow.com/questions/2701192/what-characters-can-be-used-for-up-down-triangle-arrow-without-stem-for-displa like `⭬` and `🞂`, which I did run into display issues on the win32 dashboard. So, the  `╎`  and `╼` characters do seem to be more supported than those other arrow unicode chars. 

Do you think we should include a fallback ASCII only mode? I'm not sure how to test for it (especially you're just generating a string and not necessarily sending it to a terminal stream), but we could include a flag `ascii_only=False` the user could explicitly set if they wanted. 

I think the following ascii glyphs would work: 

```
            _write("+")

            if not indent:
                if islast:
                    this_prefix = indent + "L-- "
                    next_prefix = indent + "    "
                else:
                    this_prefix = indent + "|-- "
                    next_prefix = indent + ":   "

            else:
                # For individual forests distinguish between directed and
                # undirected cases
                if is_directed:
                    if islast:
                        this_prefix = indent + "|-> "
                        next_prefix = indent + "    "
                    else:
                        this_prefix = indent + "|-> "
                        next_prefix = indent + "|   "
                else:
                    if islast:
                        this_prefix = indent + "L-- "
                        next_prefix = indent + "    "
                    else:
                        this_prefix = indent + "|-- "
                        next_prefix = indent + "|   "
```

Reply:
I pushed up a commit that did something similar. It looks very ugly, but its not much overhead to add it as a feature. Here are some demo ascii trees: 

Directed: 

```
        +-- 0
        :   |-> 1
        :   |   |-> 3
        :   |   L-> 4
        :   L-> 2
        :       |-> 5
        :       L-> 6
        +── 7
            |-> 8
            |   |-> 10
            |   L-> 11
            L-> 9
                |-> 12
                L-> 13
```

and Undirected:

```
        +-- 0
        :   |-- 1
        :   |   |-- 3
        :   |   L-- 4
        :   L-- 2
        :       |-- 5
        :       L-- 6
        +── 7
            |-- 8
            |   |-- 10
            |   L-- 11
            L-- 9
                |-- 12
                L-- 13
```